lib/vnet/setup.go

```diff
@@ -31,101 +31,138 @@ import (
+				slog.DebugContext(ctx, "Error from recvTUNErr ignored in favor of processCtx.Err", "error", err)
+				return nil, trace.Wrap(context.Cause(processCtx))
+			}
+			return nil, trace.Wrap(err, "receiving TUN from admin subcommand")
+		}
 	}
 
 	appResolver := NewTCPAppResolver(appProvider)
 
-	manager, err := NewManager(&Config{
+	ns, err := newNetworkStack(&Config{
 		TUNDevice:          tun,
 		IPv6Prefix:         ipv6Prefix,
 		DNSIPv6:            dnsIPv6,
 		TCPHandlerResolver: appResolver,
 	})
 	if err != nil {
-		return nil, nil, trace.Wrap(err)
+		return nil, trace.Wrap(err)
 	}
 
-	return manager, adminCommandErrCh, nil
+	pm.AddCriticalBackgroundTask("network stack", func() error {
+		return trace.Wrap(ns.Run(processCtx))
+	})
+
+	success = true
+	return pm, nil
 }
 
-// Run is a blocking call that waits for either the admin subcommand or the VNet manager to exit and
-// makes sure the other one exits as well.
-//
-// It captures some peculiarities of running VNet that need to be handled both in tsh and Connect.
-//
-// cancel accepts a cause because when stopping manager, we need to differentiate between regular
-// cancellation (think Ctrl + C in a terminal) vs canceling because the admin subcommand exiting
-// prematurely.
-func Run(ctx context.Context, cancel context.CancelCauseFunc, manager *Manager, adminCommandErrCh <-chan error) error {
-	allErrors := make(chan error, 4)
+func newProcessManager() (*ProcessManager, context.Context) {
+	ctx, cancel := context.WithCancel(context.Background())
 	g, ctx := errgroup.WithContext(ctx)
-	g.Go(func() error {
-		// Make sure to cancel the context if manager.Run terminates for any reason.
-		defer cancel(nil)
-		err := trace.Wrap(manager.Run(ctx), "running VNet manager")
 
-		// If ctx was canceled due to a specific cause, do not put err into allErrors. The specific
-		// cause is going to be reported by another goroutine.
-		//
-		// If this goroutine added context.Canceled into the error aggregate, final callsites in the
-		// upper layers would have assumed that it was merely a context that was canceled, while the
-		// real reason would be more complex.
-		if errors.Is(err, context.Canceled) && !errors.Is(context.Cause(ctx), context.Canceled) {
-			return nil
-		}
+	return &ProcessManager{
+		g:      g,
+		cancel: cancel,
+	}, ctx
+}
 
-		allErrors <- err
-		return err
-	})
-	g.Go(func() error {
-		var adminCommandErr error
-		select {
-		case adminCommandErr = <-adminCommandErrCh:
-			// The admin command exited before the context was canceled, cancel everything and exit.
-			// This can happen if the admin subcommand crashes or gets killed.
+// ProcessManager handles background tasks needed to run VNet.
+// Its semantics are similar to an error group with a context, but it cancels the context whenever
+// any task returns prematurely, that is, a task exits while the context was not canceled.
+type ProcessManager struct {
```

Comment:
Took me a few minutes and some poking around to fully grok this, but it's not too bad once you understand it.

I'd recommend writing a few tests for it that execute arbitrary tasks which are otherwise unrelated to VNet. Had those already existed I may have learned enough just by looking at the tests rather than having to write my own code to exercise this.

Reply:
Tests added in 3bbcc2a2c37829.